app/types/type-structures.ts

```diff
@@ -0,0 +1,317 @@
+export interface Type {
+    subtype(that: Type): boolean;
+    rsubtype?(that: Type): boolean;
+    feed(types: Map<string, Type>): Type;
+    validate(): void;
+    kind: string;
+}
+
+export class TypeScope {
+    constructor(public definitions: Map<string, Type>) {
+
+    }
+
+    feed(scope: TypeScope): TypeScope {
+        // TODO: make faster?
+        const types = new Map([...this.definitions.entries()].concat([...scope.definitions.entries()]));
+
+        for (const [key, type] of this.definitions.entries()) {
+            this.definitions.set(key, type.feed(types));
+        }
+        return this;
+    }
+}
+
+export class TypeVariable implements Type {
+    kind = "Variable";
+
+    constructor(public type?: Type, public lookupName?: string, public matchName?: string) {
```

Comment:
TypeVariables are used in two contexts, type inference, and type referencing. Either can define a match name. All three optional can be defined, or none of them, things will still work.